compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::span::Span;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind},
+    cooked::TokenKind,
+    Delim,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    pub span: Span,
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+/// Applies formatting rules to the given code str, generating edits where
+/// the source code needs to be changed to comply with the format rules.
+pub fn format(code: &str) -> Vec<Edit> {
+    let tokens = concrete::ConcreteTokenIterator::new(code);
+    let mut edits = vec![];
+
+    let mut indent_level: usize = 0;
+
+    // The sliding window used is over three adjacent tokens
+    #[allow(unused_assignments)]
+    let mut one = None;
+    let mut two = None;
+    let mut three = None;
+
+    for token in tokens {
+        // Advance the token window
+        one = two;
+        two = three;
+        three = Some(token);
+
+        let mut edits_for_triple = match (&one, &two, &three) {
+            (Some(one), Some(two), Some(three)) => {
+                match one.kind {
+                    ConcreteTokenKind::Syntax(TokenKind::Open(Delim::Brace)) => indent_level += 1,
+                    ConcreteTokenKind::Syntax(TokenKind::Close(Delim::Brace)) => {
+                        indent_level = indent_level.saturating_sub(1)
+                    }
+                    ConcreteTokenKind::WhiteSpace => continue,
+                    _ => {}
+                }
+
+                if let ConcreteTokenKind::WhiteSpace = one.kind {
+                    // first token is whitespace, continue scanning
+                    continue;
+                } else if matches!(two.kind, ConcreteTokenKind::WhiteSpace) {
+                    // whitespace in the middle
+                    apply_rules(
```

Comment:
What rules do we have? Do we have tests showing each rule being applied?

Reply:
There a many rules. We have tests for each in the test.rs file.